indexing-service/src/main/java/org/apache/druid/indexing/worker/shuffle/ShuffleMetrics.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.worker.shuffle;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ShuffleMetrics
+{
+  private final AtomicReference<ConcurrentHashMap<String, PerDatasourceShuffleMetrics>> datasourceMetrics =
+      new AtomicReference<>();
+
+  public ShuffleMetrics()
+  {
+    datasourceMetrics.set(new ConcurrentHashMap<>());
+  }
+
+  public void shuffleRequested(String supervisorTaskId, long fileLength)
+  {
+    datasourceMetrics
+        .get()
+        .computeIfAbsent(supervisorTaskId, k -> new PerDatasourceShuffleMetrics()).accumulate(fileLength);
```

Comment:
it's still possible to miss an update in reporting because of race condition, right? Since the reference could be reset while the accumulation is happening.

Reply:
The race condition exists, but it should be fine because the missing update should be included in the next call to `snapshotAndReset()`. I added javadocs explaining why.

Reply:
I think this needs to use something like `AtomicReference.getAndUpdate` so that it isn't racy with the monitor/emitter? Though I'm not sure `getAndUpdate` or the similar methods are actually appropriate since they are supposed to be side-effect free, so I'm not really sure how exactly to resolve this.

Like, the potentially problematic scenario I'm thinking of is where `shuffleRequested` is called "before" `snapshotAndReset`. It seems like once `AtomicReference.get` has completed, `snapshotAndReset` can proceed, so now the shuffle monitor has the same concurrent map we are still actively updating, and it is preparing to build the metrics to emit. It seems super unlikely that it would be a problem, but unless I'm missing something it does seem possible.

Reply:
Ah, you guys are right. Will fix it.

Reply:
The problem is that any updates on the reference to `datasourceMetrics` should be synchronized with any updates on the _map itself and its values_. I could use `ConcurrentHashMap.compute()` if I didn't have to reset the reference to the map when a snapshot is taken, but I think it's needed since the map can keep growing over time otherwise. I'm not sure if there is any other way than using a big lock. I made this change, let me know if you have a better idea.

Reply:
the lock should suffice. `shuffleRequested` doesn't need to be a high throughput call.